src/types.rs

```diff
@@ -0,0 +1,131 @@
+use pyo3::{
+    prelude::*,
+    types::{PyDict, PyList},
+    PyNativeType, ToBorrowedObject, ToPyObject,
+};
+
+pub trait PyArray<'py>: Into<PyObject> + Clone + 'py {
+    fn new<T, U>(py: Python<'py>, elements: impl IntoIterator<Item = T, IntoIter = U>) -> Self
+    where
+        T: ToPyObject,
+        U: ExactSizeIterator<Item = T>;
+
+    fn py(&self) -> Python<'py>;
+
+    fn len(&self) -> usize;
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
```

Comment:
I think most (or all) of the functions in this file probably want to be marked with `#[inline]` for performance.